docs/enhancements/fep-001-api-fleet-templating.md

```diff
@@ -0,0 +1,257 @@
+
+When a template is applied to a specific device via the Device API:
+* The template variables get rendered (i.e. the labels at that point in time get used)
+* Floating tags for external references (os image, git refs, etc..) get converted to stable tags (hashes)
+* When later in time a device has labels updated, the device controller will queue the device to be re-evaulated and
+  the fleet template re-rendered. (this is the same workflow as when the labels selected by a fleet changes)
+
+### User Stories 
+
+#### Story 1: network access credentials
+As an IT administrator, I want to use site-specific network access credentials for devices in the same fleet, so that I do not have to deal with the complexity of maintaining one fleet per site.
+
+#### Story 2: site specific endpoints
+As an IT administrator, I want to provide site-specific server endpoints to devices that depend on LAN-configured edge servers.
+
+#### Story 3: site specific certificate bundles
+As an IT administrator, I want to provide site-specific certificate bundles to enable trust with the local site infrastructure.
+
+#### Story 4: injecting the device name into the configuration
+As an IT administrator, I want the name of my MicroShift instance in ACM to have the same cluster name as the underlying device's name or fingerprint, so I can easily correlate between the two.
+
+#### Story 5: site specific telemetry endpoints
+As an IT administrator, I want to provide site-specific/region-specific telemetry endpoints to devices.
+
+#### Story 6: dialing up verbosity for logging in my user application in specific devices
+As an IT administrator, I want to dial up the verbosity for logging on a specific device, so I can troubleshoot it.
+
+#### Story 7: region specific application service endpoints
+As an IT administrator, I want to provide region-specific service endpoints to my applications
+
+#### Story 8: region specific radio bands/regulations
+As a developer, I want to ensure that my devices automatically comply with local regulations when enrolled in specific regions. (i.e. configured with region-specific radio bands for WiFi, 5G, or other radio technologies)
+
+#### Story 9: region specific data sovereignty
+As a developer, I want to ensure that my devices connect to the specific regional endpoints to ensure compliance with data sovereignty.
+
+### Notes/Constraints/Caveats
+
+We would need to identify the design changes relative to templateVersions and
+if the relationship of those continues to be to the fleet or device level.
+
+### Risks and Mitigations
+
+
+## Design Details
+
+TBD
+
+### Scalability
+
+TBD
+
+### Troubleshooting
+
+Troubleshooting can be done by looking at the device status and rendered device spec. Any errors found in templates
+could be reported as part of the fleet status conditions.
+
+## Implementation History
+
+## Drawbacks
+
+## Alternatives
```

Comment:
Is another option allowing the user to generate the templated options client side by consuming our API's?

Reply:
I don't understand what you mean :D

I need more details and an example, I will capture it here.

Reply:
Often this type of templating is done client side. Given a file template if the user has access to the fleet/device apis then can consume those apis and perform the templating on the client side.

given the example below a user could generate all the possible configurations locally. I think a good example of this is exploring how terraform generates the state file.  Complex templating is done client side and verified before pushing. In some circumstances where this could be many thousands of configurations this might not be the appropriate approach but in simpler cases I think it would be pretty useful. Also we should potentally look at how terraform handles this at scale?

```
<loop syntax>
- name: contrived-config-{{ .device.name }}
  region: {{ .device.labels["region"] }}
 </loop>
 ```

Reply:
I still don't fully understand the example.

I don't know if you mean that templating should be performed client side (the agent), in such case:

* We lose the ability to freeze floating tags into hashes (for images)
* We cannot freeze git references that have templating into plain data.

Or do you mean that some external entity must take the templating and generate the device's data?

Or something else.

I guess my friday-brain is fried :D

Can I ask you to write me a fully-fledged alternative entry?

Reply:
> Or do you mean that some external entity must take the templating and generate the device's data?


the client being the cli and the user being someone who wants to push something with git ops. I will write something up

Reply:
```
Use case: as a developer, I would like to be able to render local template config files before I push them to my GitOps workflow. This is the same process we use when we generate our Teraform state. I would expect an SDK or go client code be available to generate this from the flightctl service apis.
```

Reply:
This might play into the unmanaged usecase. I see solutions like terraform which already perform this templating efficiently with modules.  Now that IBM owns Hashicorp I think we should be considering how solutions like consol and terraform can work into these designs even if only in theory.

Reply:
At the end of the day, we're implementing look-up tables (LUTs) that map (<path_to_field_in_device_template>, <old_value>, <device_name>, <device_labels>) to <new_value>. The proposal outlined in this doc is generating the LUTs based on purely local information (go-template + device spec). IIUC, you're suggesting users should be able to upload their own LUTs. Alternatively, we could call a lambda functions provided by the users that would take the same params and compute the new values.

One challenge when using GitOps: Users will know all the labels (they'd be stored in git, too), so can include these in the LUTs. However, it's hard to include device.name (the device identity / fingerprint), as that is only known once the device starts for the first time and enrolls. And then you'd need to create a commit to add the new device id to git. Or users would have to query the current set of devices from the API, compute the LUT, then upload the LUT. For potentially a million devices. I don't think that's viable.

Reply:
> Or users would have to query the current set of devices from the API, compute the LUT, then upload the LUT. For potentially a million devices. I don't think that's viable.

This rendering would take place on the fleet level though so I would imagine smaller numbers :)